src/dmd/cparse.d

```diff
@@ -48,14 +48,16 @@ final class CParser(AST) : Parser!AST
         Array!structalign_t* packs;     // parallel alignment values
     }
 
-    /** C allows declaring a function with a typedef:
-     *   typedef int (myfunc)(); myfunc fun;
-     * but we need to distinguish `fun` being a function as opposed to a variable in the
-     * parse pass. This is accomplished by having a simple symbol table of typedefs
-     * where we know, by syntax, if they are function types or non-function types.
-     * funcTypeIds is the symbol table, of the identifiers of typedefs of function types.
+    /* C cannot be parsed without determining if an identifier is a type or a variable.
+     * For expressions like `(T)-3`, is it a cast or a minus expression?
+     * It also occurs with `typedef int (F)(); F fun;`
+     * but to build the AST we need to distinguish `fun` being a function as opposed to a variable.
+     * To fix, build a symbol table for the typedefs.
+     * Symbol table of typedefs indexed by Identifier cast to void*.
+     * 1. if an identifier is a typedef, then it will return a non-null Type
+     * 2. if an identifier is not a typedef, then it will return null
      */
-    AST.Identifiers funcTypeIds;  /// Identifiers in this are typedefs of function types
+    Array!(void*) typedefTab;  /// Array of AST.Type[Identifier], typedef's indexed by Identifier
```

Comment:
I would like to merge this but all this void* is troubling me. If the void* can't be eliminated we can at very least put this array inside a struct along with the utility functions and the bottom of this patch.

Reply:
The void* is there because the Array template cannot handle associative arrays.

There are only a couple places where the casts are necessary, it's not worth more than that.